aten/src/ATen/xpu/XPUEvent.h

```diff
@@ -0,0 +1,136 @@
+#pragma once
+#include <ATen/xpu/XPUContext.h>
+#include <c10/util/Optional.h>
+
+namespace at::xpu {
+
+/*
+ * XPUEvent are movable not copyable wrappers around SYCL event. XPUEvent are
+ * constructed lazily when first recorded. It has a device, and this device is
+ * acquired from the first recording stream. Later streams that record the event
+ * must match the same device.
+ */
+struct TORCH_XPU_API XPUEvent {
+  // Constructors
+  XPUEvent() noexcept = default;
+  XPUEvent(bool enable_timing) noexcept : enable_timing_{enable_timing} {}
+
+  ~XPUEvent() {}
+
+  XPUEvent(const XPUEvent&) = delete;
+  XPUEvent& operator=(const XPUEvent&) = delete;
+
+  XPUEvent(XPUEvent&& other) noexcept {
+    moveHelper(std::move(other));
+  }
+  XPUEvent& operator=(XPUEvent&& other) noexcept {
+    if (this != &other) {
+      moveHelper(std::move(other));
+    }
+    return *this;
+  }
+
+  operator sycl::event&() const {
+    return event();
+  }
+
+  optional<at::Device> device() const {
+    if (isCreated()) {
+      return at::Device(at::kXPU, device_index_);
+    } else {
+      return {};
+    }
+  }
+
+  bool isCreated() const {
+    return (event_.get() != nullptr);
+  }
+  DeviceIndex device_index() const {
+    return device_index_;
+  }
+  sycl::event& event() const {
+    return *event_;
+  }
+
+  bool query() const {
+    using namespace sycl::info;
+    if (!isCreated()) {
+      return true;
+    }
+
+    return event().get_info<event::command_execution_status>() ==
+        event_command_status::complete;
+  }
+
+  void record() {
+    record(getCurrentXPUStream());
+  }
+
+  void recordOnce(const XPUStream& stream) {
+    if (!isCreated()) {
+      record(stream);
+    }
+  }
+
+  void record(const XPUStream& stream) {
+    if (!isCreated()) {
+      device_index_ = stream.device_index();
+    } else {
+      TORCH_CHECK(
+          device_index_ == stream.device_index(),
+          "Event device ",
+          device_index_,
+          " does not match recording stream's device ",
+          stream.device_index(),
+          ".");
+      event_.reset();
+    }
+    event_ = std::make_unique<sycl::event>(
```

Comment:
why we do not have GPUTrace? It should be device-agnostic?

Reply:
`GPUTrace` should be device-agnostic. Currently, it has some cuda-specific code.

Reply:
Can you elaborate? I didn't find any cuda-specific code. If it does, please add a TODO here.

Reply:
`GPUTrace` is CUDA-bias, curently. see https://github.com/pytorch/pytorch/blob/995f69623d6bee50f0d413c546c838e1df1c4dcd/torch/csrc/PyInterpreter.cpp#L92-L95
and there is a hook `torch.utils._cuda_trace` in `CONCRETE_TRACE_CUDA`, see 
https://github.com/pytorch/pytorch/blob/995f69623d6bee50f0d413c546c838e1df1c4dcd/torch/csrc/PyInterpreter.cpp#L21-L30.